Tests/External/Plugins/ASCompletion.Tests/Completion/ASCompleteTests.cs

```diff
@@ -0,0 +1,692 @@
+                ASContext.Context = new HaXeContext.Context(new HaXeSettings());
+
+                var sci = GetBaseScintillaControl();
+                sci.Text = "function p(?arg:{?x:";
+                sci.ConfigurationLanguage = "haxe";
+
+                var coma = ASComplete.DisambiguateComa(sci, 20, 0);
+
+                Assert.AreEqual(ComaExpression.VarDeclaration, coma);
+            }
+
+            [Test]
+            public void HaxeFunctionOptionalArgument()
+            {
+                var pluginMain = Substitute.For<PluginMain>();
+                var pluginUiMock = new PluginUIMock(pluginMain);
+                pluginMain.MenuItems.Returns(new List<System.Windows.Forms.ToolStripItem>());
+                pluginMain.Settings.Returns(new GeneralSettings());
+                pluginMain.Panel.Returns(pluginUiMock);
+                ASContext.GlobalInit(pluginMain);
+                ASContext.Context = new HaXeContext.Context(new HaXeSettings());
+
+                var sci = GetBaseScintillaControl();
+                sci.Text = "function test(arg:String, ?arg2";
+                sci.ConfigurationLanguage = "haxe";
+
+                var coma = ASComplete.DisambiguateComa(sci, 31, 0);
+
+                Assert.AreEqual(ComaExpression.FunctionDeclaration, coma);
+            }
+
+            [Test]
+            public void HaxeFunctionOptionalArgumentType()
+            {
+                var pluginMain = Substitute.For<PluginMain>();
+                var pluginUiMock = new PluginUIMock(pluginMain);
+                pluginMain.MenuItems.Returns(new List<System.Windows.Forms.ToolStripItem>());
+                pluginMain.Settings.Returns(new GeneralSettings());
+                pluginMain.Panel.Returns(pluginUiMock);
+                ASContext.GlobalInit(pluginMain);
+                ASContext.Context = new HaXeContext.Context(new HaXeSettings());
+
+                var sci = GetBaseScintillaControl();
+                sci.Text = "function test(?arg:";
+                sci.ConfigurationLanguage = "haxe";
+
+                var coma = ASComplete.DisambiguateComa(sci, 19, 0);
+
+                Assert.AreEqual(ComaExpression.FunctionDeclaration, coma);
+            }
+
+            [Test]
+            public void HaxeTernaryOperatorTruePart()
+            {
+                var pluginMain = Substitute.For<PluginMain>();
+                var pluginUiMock = new PluginUIMock(pluginMain);
+                pluginMain.MenuItems.Returns(new List<System.Windows.Forms.ToolStripItem>());
+                pluginMain.Settings.Returns(new GeneralSettings());
+                pluginMain.Panel.Returns(pluginUiMock);
+                ASContext.GlobalInit(pluginMain);
+                ASContext.Context = new HaXeContext.Context(new HaXeSettings());
+
+                var sci = GetBaseScintillaControl();
+                sci.Text = "var a:String = (2 > 3) ?";
+                sci.ConfigurationLanguage = "haxe";
```

Comment:
Maybe we should have some enum at ASComplete level defining if we're in AS3, Haxe, etc. mode.